core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -653,22 +669,27 @@ public ValueReader<?> reader(int pos) {
     public S read(Decoder decoder, Object reuse) throws IOException {
       S struct = reuseOrCreate(reuse);
 
+      // Set the default values first. Setting default values first allows them to be overridden
+      // once the data is read from the file if the corresponding field is present in the file.
+      for (int i = 0; i < defaultValuesPositions.length; i += 1) {
```

Comment:
This can move into the `ResolvingDecoder` control flow case below.

The `else` case (not a `ResolvingDecoder`) indicates that the Avro file schema matches the expected schema because there is no expected schema. That way we know that all of the fields are read from the file and no initial defaults need to be applied..

Reply:
Done.